pkg/piecestore/rpc/client/readerwriter.go

```diff
@@ -0,0 +1,70 @@
+// Copyright (C) 2018 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package client
+
+import (
+	"fmt"
+	"log"
+
+	pb "storj.io/storj/protos/piecestore"
+)
+
+// StreamWriter -- Struct for writing piece to server upload stream
+type StreamWriter struct {
+	stream pb.PieceStoreRoutes_StoreClient
+}
+
+// Write -- Write method for piece upload to stream
+func (s *StreamWriter) Write(b []byte) (int, error) {
+	if err := s.stream.Send(&pb.PieceStore{Content: b}); err != nil {
+		return 0, fmt.Errorf("%v.Send() = %v", s.stream, err)
+	}
+
+	return len(b), nil
+}
+
+// Close -- Close Write Stream
+func (s *StreamWriter) Close() error {
+	reply, err := s.stream.CloseAndRecv()
+	if err != nil {
+		return err
+	}
+
+	log.Printf("Route summary: %v", reply)
+
+	return nil
+}
+
+// StreamReader -- Struct for reading piece download stream from server
+type StreamReader struct {
+	stream       pb.PieceStoreRoutes_RetrieveClient
+	overflowData []byte
+}
+
+// Read -- Read method for piece download stream
+func (s *StreamReader) Read(b []byte) (int, error) {
+	if len(s.overflowData) > 0 {
+		n := copy(b, s.overflowData)
+		s.overflowData = s.overflowData[:len(s.overflowData)-n]
+		return n, nil
+	}
+
+	pieceData, err := s.stream.Recv()
+	if err != nil {
+		return 0, err
+	}
+
+	n := copy(b, pieceData.Content)
+
+	if n < len(pieceData.Content) {
+		s.overflowData = b[len(b):]
```

Comment:
here's where part of the garbage collection thing comes in (though Recv kinda screws it up by not letting you pass in your own buffer). you could allocate s.overflowData once and then copy the leftover data in and that way keep less on the heap. kind of hard to explain what's happening with the allocator here without a whiteboard so remind me and i'll show you next week. nothing to change

Reply:
Honestly I am getting unhappy not being able to explicitly put a variable on the stack vs heap while coding in go. I usually do almost everything on the stack

Reply:
oh! so actually, go puts everything on the stack by default, unless the escape analysis concludes it might persist beyond the stack lifetime. this is a pretty good article about how to keep stuff on the stack and figure out when the escape analysis concludes it's impossible: http://www.agardner.me/golang/garbage/collection/gc/escape/analysis/2015/10/18/go-escape-analysis.html

one of the great things about go's named returned values is you can actually return quite a bit of data without ever using a heap allocation. sha256.Sum256 uses this trick to avoid all memory allocations: the return value is a byte array (and not a slice) and is stack allocated: https://golang.org/pkg/crypto/sha256/#Sum256

in the case here though, s lives beyond the stack lifetime so everything attached to s will be heap allocated no matter what